File: Documentospdf/ordenQuirurgica.js
// ordenQuirurgica.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { getProfessionalData } from "./professionalResolver.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export function generarOrdenQuirurgica(doc, datos) {
  const {
    nombre,
    rut,
    edad,
    diagnostico,
    codigoCirugia,
    tipoCirugia,
    modalidad, // "PAD", "Institucional", "Convencional", "Otros"
    equipoMedico,
    insumos,
    professional
  } = datos || {};

  const medico = getProfessionalData(professional);
  if (!medico) {
    throw new Error("Profesional no encontrado");
  }

  const fechaActual = new Date().toLocaleDateString("es-CL");

  /* ================= ENCABEZADO ================= */
  try {
    const logoPath = path.join(__dirname, "assets", "ica.jpg");
    if (fs.existsSync(logoPath)) {
      doc.image(logoPath, 50, 40, { width: 120 });
    }
  } catch {}

  doc.moveDown(1.5);
  doc.font("Helvetica-Bold").fontSize(18)
    .text("INSTITUTO DE CIRUGÍA ARTICULAR", 180, 50);

  doc.moveDown(1.5);
  doc.fontSize(16)
    .text("ORDEN MÉDICA DE INTERVENCIÓN QUIRÚRGICA", 150, undefined, {
      underline: true,
    });

  doc.moveDown(2);

  doc.font("Helvetica").fontSize(11)
    .text("IMPORTANTE:", { continued: true })
    .font("Helvetica")
    .text(" Se tomará contacto a la brevedad para confirmar o reagendar fecha de intervención desde la Unidad de Planificación Quirúrgica.");

  doc.moveDown(1.5);
  doc.text(`Fecha actual: ${fechaActual}`);

  doc.moveDown(2);

  /* ================= DATOS PACIENTE ================= */
  doc.font("Helvetica-Bold").fontSize(14).text("DATOS DEL PACIENTE");
  doc.moveDown(1);

  doc.font("Helvetica").fontSize(12);
  doc.text(`Nombre: ${nombre ?? ""}`);
  doc.text(`RUT: ${rut ?? ""}`);
  doc.text(`Edad: ${edad ?? ""}`);
  doc.text(`Diagnóstico: ${diagnostico ?? ""}`);
  doc.text(`Código Cirugía: ${codigoCirugia ?? ""}`);

  doc.moveDown(2);

  /* ================= DATOS INTERVENCIÓN ================= */
  doc.font("Helvetica-Bold").fontSize(14).text("DATOS DE LA INTERVENCIÓN");
  doc.moveDown(1);

  doc.font("Helvetica").fontSize(12);
  doc.text(`Tipo de Cirugía: ${tipoCirugia ?? ""}`);
  doc.moveDown(0.5);

  doc.text(`Modalidad: ${modalidad ?? ""}`);

  doc.moveDown(1);

  /* ================= EQUIPO / INSUMOS ================= */
  if (equipoMedico) {
    doc.font("Helvetica-Bold").text("Equipo Médico:");
    doc.moveDown(0.5);
    doc.font("Helvetica").text(equipoMedico);
    doc.moveDown(1);
  }

  if (insumos) {
    doc.font("Helvetica-Bold").text("Insumos / OTS:");
    doc.moveDown(0.5);
    doc.font("Helvetica").text(insumos);
  }

  /* ================= FIRMA DINÁMICA ================= */
  const pageW = doc.page.width;
  const pageH = doc.page.height;
  const marginL = doc.page.margins.left || 50;
  const marginR = doc.page.margins.right || 50;
  const baseY = pageH - 170;

  doc.font("Helvetica").fontSize(12);
  doc.text("_________________________", marginL, baseY, {
    align: "center",
    width: pageW - marginL - marginR,
  });

  doc.text("Firma Médico Tratante", marginL, baseY + 18, {
    align: "center",
    width: pageW - marginL - marginR,
  });

  // Firma
  try {
    const firmaPath = path.join(__dirname, "assets", medico.firma);
    if (fs.existsSync(firmaPath)) {
      const firmaW = 250;
      const firmaX = (pageW - firmaW) / 2;
      const firmaY = baseY - 45;
      doc.image(firmaPath, firmaX, firmaY, { width: firmaW });
    }
  } catch {}

  // Timbre
  try {
    const timbrePath = path.join(__dirname, "assets", medico.timbre);
    if (fs.existsSync(timbrePath)) {
      const firmaW = 250;
      const firmaX = (pageW - firmaW) / 2;
      const timbreW = 110;
      const timbreX = firmaX + firmaW;
      const timbreY = baseY - 65;

      doc.save();
      doc.rotate(20, {
        origin: [timbreX + timbreW / 2, timbreY + timbreW / 2],
      });
      doc.image(timbrePath, timbreX, timbreY, { width: timbreW });
      doc.restore();
    }
  } catch {}

  /* ================= PIE DINÁMICO ================= */
  doc.font("Helvetica").fontSize(12);
  doc.text(medico.nombre, marginL, baseY + 52, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text(`RUT: ${medico.rut}`, {
    align: "center",
    width: pageW - marginL - marginR,
  });
  doc.text(medico.especialidad, {
    align: "center",
    width: pageW - marginL - marginR,
  });
}

File: Documentospdf/ordenQuirurgica.py
# Documentospdf/ordenQuirurgica.py

import os
from io import BytesIO
from datetime import datetime

from reportlab.platypus import (
    SimpleDocTemplate,
    Paragraph,
    Spacer,
    Image
)
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib.pagesizes import A4

from Documentospdf.professionalResolver import getProfessionalData


def generarOrdenQuirurgica(datos: dict):

    buffer = BytesIO()

    doc = SimpleDocTemplate(
        buffer,
        pagesize=A4,
        rightMargin=50,
        leftMargin=50,
        topMargin=80,
        bottomMargin=80
    )

    elements = []
    styles = getSampleStyleSheet()

    nombre = datos.get("nombre")
    rut = datos.get("rut")
    edad = datos.get("edad")
    diagnostico = datos.get("diagnostico")
    codigoCirugia = datos.get("codigoCirugia")
    tipoCirugia = datos.get("tipoCirugia")
    modalidad = datos.get("modalidad")
    equipoMedico = datos.get("equipoMedico")
    insumos = datos.get("insumos")
    professional = datos.get("professional")

    medico = getProfessionalData(professional)
    if not medico:
        raise Exception("Profesional no encontrado")

    fechaActual = datetime.now().strftime("%d-%m-%Y")

    assets_dir = os.path.join(os.path.dirname(__file__), "assets")

    # ================= ENCABEZADO =================
    logo_path = os.path.join(assets_dir, "ica.jpg")
    if os.path.exists(logo_path):
        elements.append(Image(logo_path, width=120, height=50))
        elements.append(Spacer(1, 12))

    elements.append(Paragraph(
        "<b>INSTITUTO DE CIRUGÍA ARTICULAR</b>",
        styles["Heading1"]
    ))
    elements.append(Spacer(1, 6))

    elements.append(Paragraph(
        "<u>ORDEN MÉDICA DE INTERVENCIÓN QUIRÚRGICA</u>",
        styles["Heading2"]
    ))
    elements.append(Spacer(1, 16))

    elements.append(Paragraph(
        "<b>IMPORTANTE:</b> Se tomará contacto a la brevedad para confirmar o reagendar fecha de intervención desde la Unidad de Planificación Quirúrgica.",
        styles["Normal"]
    ))

    elements.append(Spacer(1, 12))
    elements.append(Paragraph(
        f"Fecha actual: {fechaActual}",
        styles["Normal"]
    ))
    elements.append(Spacer(1, 24))

    # ================= DATOS PACIENTE =================
    elements.append(Paragraph("<b>DATOS DEL PACIENTE</b>", styles["Heading3"]))
    elements.append(Spacer(1, 12))

    elements.append(Paragraph(f"Nombre: {nombre or ''}", styles["Normal"]))
    elements.append(Paragraph(f"RUT: {rut or ''}", styles["Normal"]))
    elements.append(Paragraph(f"Edad: {edad or ''}", styles["Normal"]))
    elements.append(Paragraph(f"Diagnóstico: {diagnostico or ''}", styles["Normal"]))
    elements.append(Paragraph(f"Código Cirugía: {codigoCirugia or ''}", styles["Normal"]))

    elements.append(Spacer(1, 24))

    # ================= DATOS INTERVENCIÓN =================
    elements.append(Paragraph("<b>DATOS DE LA INTERVENCIÓN</b>", styles["Heading3"]))
    elements.append(Spacer(1, 12))

    elements.append(Paragraph(f"Tipo de Cirugía: {tipoCirugia or ''}", styles["Normal"]))
    elements.append(Spacer(1, 6))
    elements.append(Paragraph(f"Modalidad: {modalidad or ''}", styles["Normal"]))

    elements.append(Spacer(1, 18))

    # ================= EQUIPO / INSUMOS =================
    if equipoMedico:
        elements.append(Paragraph("<b>Equipo Médico:</b>", styles["Normal"]))
        elements.append(Spacer(1, 6))
        elements.append(Paragraph(equipoMedico, styles["Normal"]))
        elements.append(Spacer(1, 12))

    if insumos:
        elements.append(Paragraph("<b>Insumos / OTS:</b>", styles["Normal"]))
        elements.append(Spacer(1, 6))
        elements.append(Paragraph(insumos, styles["Normal"]))
        elements.append(Spacer(1, 12))

    elements.append(Spacer(1, 40))

    # ================= FIRMA =================
    elements.append(Paragraph("______________________________", styles["Normal"]))
    elements.append(Spacer(1, 6))
    elements.append(Paragraph("Firma Médico Tratante", styles["Normal"]))
    elements.append(Spacer(1, 12))

    # Firma
    firma_path = os.path.join(assets_dir, medico.get("firma", ""))
    if os.path.exists(firma_path):
        elements.append(Image(firma_path, width=200, height=60))
        elements.append(Spacer(1, 12))

    # Timbre
    timbre_path = os.path.join(assets_dir, medico.get("timbre", ""))
    if os.path.exists(timbre_path):
        elements.append(Image(timbre_path, width=100, height=100))
        elements.append(Spacer(1, 12))

    # ================= PIE =================
    elements.append(Spacer(1, 12))
    elements.append(Paragraph(f"<b>{medico.get('nombre','')}</b>", styles["Normal"]))
    elements.append(Paragraph(f"RUT: {medico.get('rut','')}", styles["Normal"]))
    elements.append(Paragraph(medico.get("especialidad",""), styles["Normal"]))

    doc.build(elements)

    buffer.seek(0)
    return buffer
